azerothcore/azerothcore-wotlk
# feat(Core/Hooks): Functions to improve core-based modding support

Added two new hooks, BeforeCastTime that runs after a spell is prepared but before begging it's cast time and WhileCast that runs every time the update function for a spell runs while casting or channeling, as long as m_timer does not become 0. 

Added ModifySpellValue for more easily modifying spell values more easily during more awkward points in a spells lifetime.

Added an std::map<MapDummy, std::optional<std::any>> triggerDummy; where MapDummy is a class enum, to allow modded spells to have optional data assigned to them for more complex spells, especially ones that happen from interactions between different spells, such as using one spell to cast another, useful for spells that rely on one or more dummy spells that use more complex input from one "phase" to another.

## Changes Proposed:
-  Add two new hooks, BeforeCastTime and WhileCast.
-  Added function ModifySpellValue 
- Added std::map<MapDummy, std::optional<std::any>> to easily track data from one spell to another.

## Issues Addressed:
None

## SOURCE:
I made it up

## Tests Performed:
- Builds and runs without errors.
- Example of spell that utilizes new hooks to tell if a player is in the air while casting the spell, delaying the cast time until they land and gaining damage based on how long they were in the air. https://streamable.com/5q7arz


## How to Test the Changes:
1. Create a spell script that on each event of the new hooks casts a spell with triggered = true and cast it.
2. Create a spell script that modifies effect values on hook using ModifySpellValue and cast it.
3. Create a spell called BetterPyroblast that casts Pyroblast using hooks and use FindCurrentSpellBySpellId to find the Pyroblast, and an enum in MapDummy called CastByBetterPyroblast. Use GetTriggerDummy to get the reference to triggerDummy and set triggerDummy[MapDummy::CastByBetterPyroblast] = true. Create an on-hit hook for Pyroblast that detects if triggerDummy[MapDummy::CastByBetterPyroblast] has a value and check if it's equal to true, casts a third dummy stun on the target.

## Known Issues and TODO List:
- None


## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.
